### PR TITLE
Ability to set a default environment at launch

### PR DESCRIPTION
**Pull request by electric-souperman**
#### Reasons for the change (please link any relevant issues!)
* #12608 requested the option to set a `default_start_environment` variable in `.condarc` which would allow the user to default to their most regularly used env on launch, rather than base (which is the current functionality).
* I spoke with @jezdez about this, who indicated this will require a design decision before being taken forward (hence the removal of the good first issue tag).
* At @jezdez suggestion, this PR is offered in the spirit of aiding discussion about whether this functionality should be implemented, by demonstrating how it might be implemented.
#### Solution
* I have added a new primitive parameter, `default_start_environment`, to the `Context` class in `conda/base/context.py`, with default value of `"base"`.
* I have modified the way `_Activator.env_name_or_prefix` is initialized in `conda/activate.py` so that if the `activate` command is not accompanied with an env name, the `env_name_or_prefix` variable defaults to the value of `context.default_start_environment`.

#### Testing
I have implemented two unit tests to confirm that;
1. setting the `default_start_environment` in `.condarc` actually changes which env `conda activate` activates, and 
2. not setting `default_start_environment` means that `conda activate` activates base.

The changes in this PR pass both these tests.

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

**Comment by electric-souperman** [comment]
> We require contributors to sign our [Contributor License Agreement](https://conda.io/en/latest/contributing.html#conda-contributor-license-agreement) and we don't have one on file for @electric-souperman.
> 
> In order for us to review and merge your code, please e-sign the [Contributor License Agreement PDF](https://conda.io/en/latest/contributing.html#conda-contributor-license-agreement). We then need to manually verify your signature. We will ping the bot to refresh the PR status when we have confirmed your signature.

Can confirm I have signed the CLA, but it was only a few minutes prior to submitting the PR so perhaps the bot missed it. I re-signed with the @ at the beginning of my github ID in case that was the problem.

File: docs/source/user-guide/configuration/use-condarc.rst [hunk 1]
```diff
@@ -691,6 +692,24 @@ EXAMPLE:
    This is forced to ``True`` if conda-build is installed and older than 3.18.3,
    because older versions of conda break when conda feeds it the new file format.
 
+.. _set-default-env:
+Specify a default environment to activate on launch
+---------------------------------------------------
+It is now possible to set a default conda environment at launch, so when you open 
+an Anaconda Prompt, or enter ``conda activate`` at the command line, it will activate 
+your preferred environment rather than ``base``. This is done by setting the 
+``default_start_environment`` variable.
```
Comment (travishathaway) [inline docs/source/user-guide/configuration/use-condarc.rst:701]:
```suggestion
Conda 23.5.0 introduced the ability to set a default environment on shell start-up. This is done
by setting the ``default_start_environment`` variable.
```

Comment (travishathaway) [reply docs/source/user-guide/configuration/use-condarc.rst:701]: @electric-souperman, we now avoid referencing Anaconda within the conda documentation. Additionally, our users can use more than just the Anaconda prompt to run conda commands, so the above may be confusing to those users.

File: news/12608-ability-to-set-default-env [hunk 2]
```diff
@@ -0,0 +1,21 @@
+### Enhancements
+
+* Ability to set a default conda environment at launch, so when you open an Anaconda Prompt, or enter `conda activate` at the command line, it will activate your preferred environment rather than `base`.
```
Comment (travishathaway) [inline news/12608-ability-to-set-default-env:3]:
```suggestion
* Ability to set a default conda environment at launch by introducing a new configuration variable (see documentation for more details)
```

Again, I am removing the reference to Anaconda Prompt.

File: news/12608-ability-to-set-default-env [hunk 3]
```diff
@@ -0,0 +1,21 @@
+### Enhancements
+
+* Ability to set a default conda environment at launch, so when you open an Anaconda Prompt, or enter `conda activate` at the command line, it will activate your preferred environment rather than `base`. 
+* The command to set the default conda environment is `conda config --set default_start_environment my-favourite-env`.
+* If you don't set a default start environment, `conda activate` will continue to activate `base`.
```
Comment (travishathaway) [inline news/12608-ability-to-set-default-env:5]:
```suggestion
```

It's better just have a single line per section. There's already plenty of information in the documentation, so no need to have so much here.

File: news/12608-ability-to-set-default-env [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+* <news item>
+
+### Deprecations
+
+* <news item>
+
+### Docs
+
+* Updated `docs/user-guide/configuration/use-condarc.rst` to reflect this change.
```
Comment (travishathaway) [inline news/12608-ability-to-set-default-env:17]:
```suggestion
* Updated `docs/user-guide/configuration/use-condarc.rst` to describe the usage of a new configuration variable `default_start_environment`
```

When these items get complied into our release document, they will not be next to each other. So, it's better to be more specific when describing what is done. See this release document to get a better understanding of what I am talking about: https://github.com/conda/conda/releases/tag/23.3.0

**Comment by beeankha** [comment]
@conda-bot check

File: docs/source/user-guide/configuration/use-condarc.rst [hunk 5]
```diff
@@ -691,6 +692,22 @@ EXAMPLE:
...
+by setting the ``default_start_environment`` variable.
+
+EXAMPLE:
+
+.. code-block:: yaml
+
+  default_start_environment: my-favourite-env
+
+.. note::
+  If ``default_start_environment`` is not set (or is set and subsequently removed), conda
```
Comment (jaimergp) [inline docs/source/user-guide/configuration/use-condarc.rst:708]:
```suggestion
  If ``auto_activate_environment`` is not set (or is set and subsequently removed), conda 
```

File: news/12608-ability-to-set-default-env [hunk 6]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* Ability to set a default conda environment at launch by introducing a new configuration variable (see documentation for more details)
```
Comment (jaimergp) [inline news/12608-ability-to-set-default-env:3]:
```suggestion
* Ability to auto-activate a non-base conda environment at launch via the `auto_activate_environment` setting. (#12618)
```

File: news/12608-ability-to-set-default-env [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+* <news item>
+
+### Deprecations
+
+* <news item>
+
+### Docs
+
+* Updated `docs/user-guide/configuration/use-condarc.rst` to describe the usage of a new configuration variable `default_start_environment`
```
Comment (jaimergp) [inline news/12608-ability-to-set-default-env:15]:
```suggestion
* Updated `docs/user-guide/configuration/use-condarc.rst` to describe the usage of a new configuration variable `auto_activate_environment`. (#12618)
```

File: tests/condarc [hunk 8]
```diff
@@ -63,3 +63,7 @@ disallow:
 # enable certain features to be tracked by default
 track_features:
   - mkl
+
+# set an environment to default to instead of base
+default_start_environment:
```
Comment (jaimergp) [inline tests/condarc:68]:
```suggestion
auto_activate_environment:
```

File: tests/test_activate.py [hunk 9]
```diff
@@ -1207,6 +1210,31 @@ def test_build_activate_restore_unset_env_vars(self):
                     assert builder["activate_scripts"] == ()
                     assert builder["deactivate_scripts"] == ()
 
+    def test_set_default_start_env(self):
+        # create an env called test-env, and then set it to the default
+        exit_code_1 = main_subshell("create", "-y", "-n", "test-env")
+        exit_code_2 = main_subshell("config", "--set", "default_start_environment", "test-env")
```
Comment (jaimergp) [inline tests/test_activate.py:1216]:
```suggestion
        exit_code_2 = main_subshell("config", "--set", "auto_activate_environment", "test-env")
```

**Review by jaimergp — CHANGES_REQUESTED** [review]
Thanks for the changes!

As discussed earlier, I strongly favor a `auto_activate_environment` name for this variable, to keep it harmonized with the previously existing `auto_activate_base`.

The docs for the latter should be updated to favor the new version (e.g. `auto_activate_base: true` becomes `auto_activate_environment: base`; `auto_activate_base: false` becomes `auto_activate_environment: null`). The `context` attribute should also be marked for deprecation. A test should be included to clarify what happens when both keys are set (e.g. which one takes precedence).

**Comment by jaimergp** [comment]
pre-commit.ci autofix

File: docs/source/user-guide/configuration/use-condarc.rst [hunk 10]
```diff
@@ -691,6 +692,22 @@ EXAMPLE:
    This is forced to ``True`` if conda-build is installed and older than 3.18.3,
    because older versions of conda break when conda feeds it the new file format.
 
+.. _set-default-env:
+Specify a default environment to activate on launch
+---------------------------------------------------
+Conda 23.5.0 introduced the ability to set a default environment on shell start-up. This is done
```
[the remark this answers is not in the document]
Comment (electric-souperman) [reply docs/source/user-guide/configuration/use-condarc.rst]: It wouldn't let me commit on here so have done so locally on my machine. Will push shortly.

File: docs/source/user-guide/configuration/use-condarc.rst (hunk 5 above)
Comment (electric-souperman) [reply docs/source/user-guide/configuration/use-condarc.rst]: It wouldn't let me commit on here so have done so locally on my machine. Will push shortly.

File: tests/test_activate.py [hunk 11]
```diff
@@ -14,6 +14,7 @@
 from tempfile import gettempdir
 from unittest import TestCase
 from uuid import uuid4
+from typing import Literal, LiteralString
```
[the remark this answers is not in the document]
Comment (electric-souperman) [reply tests/test_activate.py]: Have removed this import and references to LiteralString and the affected unit tests still work :)

File: tests/test_activate.py [hunk 12]
```diff
@@ -1227,18 +1228,22 @@ def test_set_default_start_env(self):
...
-            "config", "--remove-key", "default_start_environment"
+            "config", "--remove-key", "auto_activate_environment"
         )
         context.__init__()
 
-    def test_default_start_env_base_if_not_set(self):
+    def test_auto_activate_env_base_if_not_set(self):
         # activate an unspecified env
-        activator = PosixActivator(["activate"])
+        activator = (PowerShellActivator(["activate", "--no-stack"]) if on_win else PosixActivator(["activate", "--no-stack"]))
```
Comment (jaimergp) [inline tests/test_activate.py:1236]:
```suggestion
        activator = PowerShellActivator(["activate", "--no-stack"] if on_win else PosixActivator(["activate", "--no-stack"]))
```

File: tests/test_activate.py [hunk 13]
```diff
@@ -1227,18 +1228,22 @@ def test_set_default_start_env(self):
...
         activator.execute()
 
         # if the we've managed to activate the base environment, the test passes
         assert activator.env_name_or_prefix == "base"
 
+        # isolate the test by clearing out the env and resetting the context
+        activator.deactivate()
+        context.__init__()
+
```
Comment (jaimergp) [inline tests/test_activate.py:1245]: I think we have some helpers to ensure the context is reset properly in `conda.testing.*`. That said, I don't think `activate()` or `deactivate()` really have effects on a running Python process? The shell code needs to be eval'd in the shell, usually. What's the concern here?

**Comment by jaimergp** [comment]
While looking at other parts of the code yesterday I saw there's a `CONDA_DEFAULT_ENV` environment variable checked here and there, as well as a `context.default_env`. Maybe we can repurpose it for this? Check:

https://github.com/conda/conda/blob/bd9c963edda8499ca9f958536abbb8e8b2857f57/src/conda-shell/conda_shell/activate.py#L679

https://github.com/conda/conda/blob/bd9c963edda8499ca9f958536abbb8e8b2857f57/src/conda-shell/conda_shell/activate.py#L542-L544

https://github.com/conda/conda/blob/bd9c963edda8499ca9f958536abbb8e8b2857f57/src/libconda/conda/base/context.py#L695-L708

It seems to be more tied to the activation logic than an option (i.e. a value being passed around), but it does look like a half-baked feature in the direction of this PR.

File: setup.cfg [hunk 14]
```diff
@@ -16,10 +16,10 @@ addopts =
...
     --strict-markers
     --durations 16
-    --xdoctest-modules
-    --xdoctest-style=google
-    --cov-report xml
-    --cov-append
+    #--xdoctest-modules
+    #--xdoctest-style=google
+    #--cov-report xml
+    #--cov-append
```
Comment (travishathaway) [inline setup.cfg:22]: Please undo this 😅.

**Pull request closed without merging**